waspc/src/Wasp/Generator/JsImport.hs

```diff
@@ -1,26 +1,122 @@
+      { _importName :: EI.ExtImportName,
+        _importPath :: Path Posix (Rel Dir') File',
+        _internalImportAlias :: Maybe JsImportAlias
+      }
+  | -- | Imports based on user defined imports in .wasp file
+    UserJsImportDefinition
+      { _scope :: JsImportScope,
+        _relPathToRoot :: RelPathToRoot,
+        _extImport :: EI.ExtImport,
+        _userImportAlias :: Maybe JsImportAlias
+      }
+  deriving (Show)
+
+-- | User code is imported either on client or server side
+data JsImportScope = WebApp | Server
+  deriving (Show)
+
+-- | Relative path to directory e.g. ../../ in ../../server/src/ext-src
+type RelPathToRoot = Path Posix (Rel Dir') Dir'
+
+-- | Alias in import e.g. @import { Identifier as Alias } from "file.js"@
+type JsImportAlias = String
+
+-- | JsImportIdentifier is used to access the imported module
+--   JsImportStatement is the actual import statement
+type JsImportTemplateData = (JsImportIdentifier, JsImportStatement)
+
+-- | Generates import statements based on import definition
+getImportStatementData :: JsImportDefinition -> JsImportTemplateData
+
+-- | For internal imports
+getImportStatementData (InternaJslImportDefinition importName importPath importAlias) =
+  (importIdentifier, importStatement)
   where
-    importStmt = "import " ++ importWhat ++ " from '" ++ importFrom ++ "'"
-    importFrom = "./" ++ SP.fromRelFileP (relPosixPathToExtCodeDir </> SP.castRel (AS.ExtImport.path extImport))
-    (importIdentifier, importWhat) =
-      case AS.ExtImport.name extImport of
-        AS.ExtImport.ExtImportModule defaultImport -> (defaultImport, defaultImport)
-        AS.ExtImport.ExtImportField namedImport -> (namedImport, "{ " ++ namedImport ++ " }")
+    (importIdentifier, importWhat) = getFirstPartOfImport importName importAlias
+    importStatement = getImportStatement importWhat importPath
+-- For user defined imports
+getImportStatementData (UserJsImportDefinition scope relPathToRoot extImport importAlias) =
+  (importIdentifier, importStatement)
+  where
+    userDefinedPath = SP.castRel $ EI.path extImport
+    webAppExtDir = fromJust (SP.relDirToPosix . SP.castRel $ extClientCodeDirInWebAppSrcDir)
+    serverExtDir = fromJust (SP.relDirToPosix . SP.castRel $ extServerCodeDirInServerSrcDir)
+    importFrom =
+      case scope of
+        WebApp -> relPathToRoot </> webAppExtDir </> userDefinedPath
+        Server -> relPathToRoot </> serverExtDir </> userDefinedPath
+    (importIdentifier, importWhat) = getFirstPartOfImport (EI.name extImport) importAlias
+    importStatement = getImportStatement importWhat importFrom
+
+type JsImportStatement = String
+
+getImportStatement :: JsImportWhat -> Path Posix (Rel Dir') File' -> JsImportStatement
+getImportStatement importWhat importPath =
+  "import " ++ importWhat ++ " from '" ++ normalizedPath ++ "'"
+  where
+    filePath = SP.fromRelFileP importPath
+    -- TODO: check if StrongPath can generate "./" prefixed paths
```

Comment:
Since I've decided to use `[reldir|../|]` to indicate relative path to the `root` folder when generating the `importPath` ([line](https://github.com/wasp-lang/wasp/pull/956/files#diff-883fead5eb518073046ab034da6eb5e9918a6544075416dd8c51ef415db098e5R45)) an edge case popped up. 

When the file importing from `ext-src` is at top level e.g. `import Page from ./ext-src/Page.jsx` and we encode this like `[reldir|./|] </> webAppExtDir </> pathToFile` the `StrongPath` lib renders it like `ext-src/Page.jsx` (without the `./` prefix) which makes it an absolute import in JS 😢 

@Martinsos Is there a better way force the `./` prefix?

Reply:
Ha yes -> well I think this is ok from the side of StrongPath, because having ./ at the front doesn't really make much sense path-wise -> it is a redundancy, gives no info. So while StrongPath will maintain ../ in front, ./ is normalized away when normalizing the path.

So if you want to make sure to force it, I don't think there is a better option but to check if ./ exists at the start and add it if needed.